javascript/ql/src/semmle/javascript/dataflow/TypeTracking.qll

```diff
@@ -0,0 +1,180 @@
+    exists(Boolean hasReturn2, Boolean hasCall2 |
+      that = MkStepSummary(hasReturn2, hasCall2)
+    |
+      result = MkStepSummary(hasReturn.booleanOr(hasReturn2), hasCall.booleanOr(hasCall2)) and
+      // avoid constructing invalid paths
+      not (hasCall = true and hasReturn2 = true)
+    )
+  }
+
+  /**
+   * Gets the summary for the path obtained by appending `this` to `that`.
+   */
+  StepSummary prepend(StepSummary that) { result = that.append(this) }
+
+  /** Gets a textual representation of this path summary. */
+  string toString() {
+    exists(string withReturn, string withCall |
+      (if hasReturn = true then withReturn = "with" else withReturn = "without") and
+      (if hasCall = true then withCall = "with" else withCall = "without")
+    |
+      result = "path " + withReturn + " return steps and " + withCall + " call steps"
+    )
+  }
+}
+
+module StepSummary {
+  /**
+   * Gets a summary describing a path without any calls or returns.
+   */
+  StepSummary level() { result = MkStepSummary(false, false) }
+
+  /**
+   * Gets a summary describing a path with one or more calls, but no returns.
+   */
+  StepSummary call() { result = MkStepSummary(false, true) }
+
+  /**
+   * Gets a summary describing a path with one or more returns, but no calls.
+   */
+  StepSummary return() { result = MkStepSummary(true, false) }
+
+  /**
+   * INTERNAL: Use `SourceNode.track()` instead.
+   */
+  predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ, StepSummary summary) {
+    exists (DataFlow::Node predNode | pred.flowsTo(predNode) |
+      // Flow through properties of objects
+      propertyFlowStep(predNode, succ) and
+      summary = level()
+      or
+      // Flow through global variables
+      globalFlowStep(predNode, succ) and
+      summary = level()
+      or
+      // Flow into function
+      callStep(predNode, succ) and
+      summary = call()
+      or
+      // Flow out of function
+      returnStep(predNode, succ) and
+      summary = return()
+    )
+    or
+    fieldStep(pred, succ) and
+    summary = level()
+  }
+
+  /**
+   * INTERNAL. Do not use.
+   *
+   * Appends a step summary onto a type-tracking summary.
+   */
+  TypeTracker append(TypeTracker type, StepSummary summary) {
+    not (type.hasCall() = true and summary.hasReturn() = true) and
+    result.hasCall() = type.hasCall().booleanOr(summary.hasCall())
+  }
+}
+
+private newtype TTypeTracker = MkTypeTracker(boolean hasCall) {
+  hasCall = true or hasCall = false
+}
+
+/**
+ * Summary of the steps needed to track a value to a given dataflow node.
+ * 
+ * This can be used to track objects that implement a certain API in order to
+ * recognize calls to that API. Note that type-tracking does not provide a
+ * source/sink relation, that is, it may determine that a node has a given type,
+ * but it won't determine where that type came from.
+ *
+ * It is recommended that all uses of this type is written on the following form,
```

Comment:
s/is written/are written/

Also, I'm not sure "on the following form" is idiomatic. Perhaps "in the following way" would be better.